File: Components/WinForms/VB/SchedulerMainDemo/Modules/OutlookImport.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Windows.Forms
Imports DevExpress.XtraScheduler
Imports DevExpress.Xpo
Imports DevExpress.XtraScheduler.Exchange
Imports DevExpress.XtraScheduler.Outlook

Namespace DevExpress.XtraScheduler.Demos
	Partial Public Class OutlookImportModule
		Inherits DevExpress.XtraScheduler.Demos.TutorialControl
		Public Sub New()
			' This call is required by the Windows Form Designer.
			InitializeComponent()

			' TODO: Add any initialization after the InitializeComponent call
		End Sub

		Public Overrides ReadOnly Property PrintingSchedulerControl() As SchedulerControl
			Get
				Return schedulerControl
			End Get
		End Property
		Protected ReadOnly Property IsCancelForRecurring() As Boolean
			Get
				Return (CType(cbCancelingTypes.EditValue, UsedAppointmentType)).Equals(UsedAppointmentType.Recurring)
			End Get
		End Property
		Protected ReadOnly Property IsCancelForNonRecurring() As Boolean
			Get
				Return (CType(cbCancelingTypes.EditValue, UsedAppointmentType)).Equals(UsedAppointmentType.NonRecurring)
			End Get
		End Property
		Protected ReadOnly Property CalendarName() As String
			Get
				Return Me.cbCalendarFolderPaths.Text
			End Get
		End Property

		Private Sub OutlookImportModule_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
			schedulerControl.Start = DateTime.Today
			UpdateControls()
		End Sub
		Private Sub OutlookImportModule_VisibleChanged(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.VisibleChanged
			If Visible Then
				xpCollectionEvents.Session.DropIdentityMap()
				xpCollectionEvents.Reload()
			End If
		End Sub
		Private Sub UpdateControls()
			cbCancelingTypes.EditValue = UsedAppointmentType.None
			FillCalendarNamesCombo()
		End Sub
		Private Sub FillCalendarNamesCombo()
			cbCalendarFolderPaths.Properties.BeginUpdate()
			Try
				cbCalendarFolderPaths.Properties.Items.Clear()
				cbCalendarFolderPaths.Properties.Items.AddRange(DemoUtils.OutlookCalendarPaths)
			Finally
				cbCalendarFolderPaths.Properties.EndUpdate()
				cbCalendarFolderPaths.SelectedIndex = 0
			End Try
		End Sub
		Private Sub schedulerStorage_AppointmentsChanged(ByVal sender As Object, ByVal e As PersistentObjectsEventArgs) Handles schedulerStorage.AppointmentsInserted, schedulerStorage.AppointmentsChanged
			session1.BeginTransaction()
			Try
				For Each apt As Appointment In e.Objects
					Dim o As XPBaseObject = TryCast(apt.GetSourceObject(schedulerStorage), XPBaseObject)
					If o IsNot Nothing Then
						o.Save()
					End If
				Next apt
				session1.CommitTransaction()
			Catch
				session1.RollbackTransaction()
			End Try
		End Sub

		#Region "supplementary functionality"
		Private Sub AddToLog(ByVal s As String)
			Me.lbcLog.Items.Add(s)
		End Sub
		Private Sub SuspendLog()
			Me.lbcLog.Items.BeginUpdate()
		End Sub
		Private Sub ResumeLog()
			Me.lbcLog.SelectedIndex = Me.lbcLog.Items.Count - 1
			Me.lbcLog.Items.EndUpdate()
		End Sub
		Private Sub ClearLog()
			Me.lbcLog.Items.Clear()
		End Sub
		Private Sub BeforeImport(ByVal objectCount As Integer)
			AddToLog(String.Format("{0} - Start importing from '{1}' calendar...", DateTime.Now.ToLongTimeString(), CalendarName))
			Application.DoEvents()
			SuspendLog()
			ProgressBarControl1.Position = 0
			ProgressBarControl1.Properties.Maximum = objectCount
		End Sub
		Private Sub AfterImport()
			AddToLog(String.Format("{0} - Import complete!", DateTime.Now.ToLongTimeString()))
			AddToLog(New String("-"c, 60))
			ResumeLog()
			ProgressBarControl1.Properties.Maximum = 100
			ProgressBarControl1.Position = 0
		End Sub
		Private Function CreateImportLogString(ByVal e As AppointmentImportingEventArgs) As String
			Dim args As OutlookAppointmentImportingEventArgs = CType(e, OutlookAppointmentImportingEventArgs)
			Return String.Format("Trying to import: apt={0}, olApt={1} ...", DemoUtils.FormatAppointmentString(args.Appointment), DemoUtils.FormatOutlookAppointmentString(args.OutlookAppointment))
		End Function
		#End Region

		Private Sub btnClearLog_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnClearLog.Click
			ClearLog()

		End Sub
		Private Sub btnImport_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnImport.Click
			Try
				Import()
			Catch ex As Exception
				DemoUtils.ReportOutlookError("import appointments from Microsoft Outlook", ex.Message)
			End Try
		End Sub
		Private Sub Import()
			Dim importer As AppointmentImporter = schedulerStorage.CreateOutlookImporter()
			CType(importer, ISupportCalendarFolders).CalendarFolderName = CalendarName
			AddHandler importer.AppointmentImporting, AddressOf importer_AppointmentImporting
			AddHandler importer.AppointmentImported, AddressOf importer_AppointmentImported
			AddHandler importer.OnException, AddressOf importer_OnException

			Dim oldCur As Cursor = Cursor.Current
			Cursor.Current = Cursors.WaitCursor
			BeforeImport(importer.SourceObjectCount)
			Try
				importer.Import(System.IO.Stream.Null)
			Finally
				AfterImport()
				Cursor.Current = oldCur
			End Try
		End Sub

		Private Sub importer_OnException(ByVal sender As Object, ByVal e As ExchangeExceptionEventArgs)
			Dim errText As String = e.OriginalException.Message
			Dim canTerminate As Boolean = True
			Dim occurrenceArgs As OutlookExchangeChangedOccurrenceExceptionEventArgs = TryCast(e, OutlookExchangeChangedOccurrenceExceptionEventArgs)
			If occurrenceArgs IsNot Nothing Then
				canTerminate = False
				If occurrenceArgs.OutlookAppointment IsNot Nothing AndAlso occurrenceArgs.Occurrence IsNot Nothing Then
					errText &= String.Format(Constants.vbCrLf & "Changed occurrence with index {0} is imported as normal occurrence '{1}' started on {2:D} at {3}" & Constants.vbLf, occurrenceArgs.Occurrence.RecurrenceIndex, occurrenceArgs.Occurrence.Subject, occurrenceArgs.Occurrence.Start, occurrenceArgs.Occurrence.Start.TimeOfDay)
				End If
			Else
				Dim args As OutlookExchangeExceptionEventArgs = TryCast(e, OutlookExchangeExceptionEventArgs)
				If args IsNot Nothing Then
					If args.OutlookAppointment IsNot Nothing Then
						errText &= String.Format(Constants.vbCrLf & "Event '{0}' started on {1:D} at {2}" & Constants.vbLf, args.OutlookAppointment.Subject, args.OutlookAppointment.Start, args.OutlookAppointment.Start.TimeOfDay)
					End If
				End If
			End If

			AddToLog(errText)
			Dim importer As AppointmentImporter = CType(sender, AppointmentImporter)

			e.Handled = True
			If canTerminate Then
				importer.Terminate()
				Throw e.OriginalException
			End If
		End Sub
		Private Sub importer_AppointmentImporting(ByVal sender As Object, ByVal e As AppointmentImportingEventArgs)
			Dim args As OutlookAppointmentImportingEventArgs = CType(e, OutlookAppointmentImportingEventArgs)
			AddToLog(CreateImportLogString(args))

			ProgressBarControl1.Position += 1
			ProgressBarControl1.Update()

			Dim cancel As Boolean = If(args.OutlookAppointment.IsRecurring, IsCancelForRecurring, IsCancelForNonRecurring)
			If cancel Then
				e.Cancel = True
				AddToLog("Cancelled because of the appointment type")
			End If
		End Sub
		Private Sub importer_AppointmentImported(ByVal sender As Object, ByVal e As AppointmentImportedEventArgs)
			AddToLog(String.Format("Successfully imported: {0}", DemoUtils.FormatAppointmentString(e.Appointment)))
		End Sub
	End Class
End Namespace

File: Components/WinForms/CS/SchedulerMainDemo/Modules/OutlookImport.cs
using System;
using System.Windows.Forms;
using DevExpress.XtraScheduler;
using DevExpress.Xpo;
using DevExpress.XtraScheduler.Exchange;
using DevExpress.XtraScheduler.Outlook;

namespace DevExpress.XtraScheduler.Demos {
	public partial class OutlookImportModule : DevExpress.XtraScheduler.Demos.TutorialControl {
		public OutlookImportModule() {
			// This call is required by the Windows Form Designer.
			InitializeComponent();

			// TODO: Add any initialization after the InitializeComponent call
		}

		public override SchedulerControl PrintingSchedulerControl { get { return schedulerControl; } }
		protected bool IsCancelForRecurring { get { return ((UsedAppointmentType)cbCancelingTypes.EditValue).Equals(UsedAppointmentType.Recurring); } }
		protected bool IsCancelForNonRecurring { get { return ((UsedAppointmentType)cbCancelingTypes.EditValue).Equals(UsedAppointmentType.NonRecurring); } }
        protected string CalendarName { get { return this.cbCalendarFolderPaths.Text; } }

		private void OutlookImportModule_Load(object sender, EventArgs e) {
			schedulerControl.Start = DateTime.Today;
			UpdateControls();
		}
		private void OutlookImportModule_VisibleChanged(object sender, EventArgs e) {
			if (Visible) {
				xpCollectionEvents.Session.DropIdentityMap();
				xpCollectionEvents.Reload();
			}
		}
		private void UpdateControls() {
			cbCancelingTypes.EditValue = UsedAppointmentType.None;
            FillCalendarNamesCombo();
		}
        private void FillCalendarNamesCombo() {
            cbCalendarFolderPaths.Properties.BeginUpdate();
            try {
                cbCalendarFolderPaths.Properties.Items.Clear();
				cbCalendarFolderPaths.Properties.Items.AddRange(DemoUtils.OutlookCalendarPaths);
            } finally {
                cbCalendarFolderPaths.Properties.EndUpdate();
                cbCalendarFolderPaths.SelectedIndex = 0;
            }
        }
		private void schedulerStorage_AppointmentsChanged(object sender, PersistentObjectsEventArgs e) {
			session1.BeginTransaction();
			try {
				foreach (Appointment apt in e.Objects) {
					XPBaseObject o = apt.GetSourceObject(schedulerStorage) as XPBaseObject;
					if (o != null)
						o.Save();
				}
				session1.CommitTransaction();
			}
			catch {
				session1.RollbackTransaction();
			}
		}

		#region supplementary functionality
		void AddToLog(string s) {
			this.lbcLog.Items.Add(s);
		}
		void SuspendLog() {
			this.lbcLog.Items.BeginUpdate();
		}
		void ResumeLog() {
			this.lbcLog.SelectedIndex = this.lbcLog.Items.Count - 1;
			this.lbcLog.Items.EndUpdate();
		}
		void ClearLog() {
			this.lbcLog.Items.Clear();
		}
		void BeforeImport(int objectCount) {
            AddToLog(String.Format("{0} - Start importing from '{1}' calendar...", DateTime.Now.ToLongTimeString(), CalendarName));
			Application.DoEvents();
			SuspendLog();
			ProgressBarControl1.Position = 0;
			ProgressBarControl1.Properties.Maximum = objectCount;
		}
		void AfterImport() {
            AddToLog(String.Format("{0} - Import complete!", DateTime.Now.ToLongTimeString()));
			AddToLog(new string('-', 60));
			ResumeLog();
			ProgressBarControl1.Properties.Maximum = 100;
			ProgressBarControl1.Position = 0;
		}
		string CreateImportLogString(AppointmentImportingEventArgs e) {
			OutlookAppointmentImportingEventArgs args = (OutlookAppointmentImportingEventArgs)e;
			return String.Format("Trying to import: apt={0}, olApt={1} ...",
				DemoUtils.FormatAppointmentString(args.Appointment), DemoUtils.FormatOutlookAppointmentString(args.OutlookAppointment));
		}
		#endregion

		private void btnClearLog_Click(object sender, EventArgs e) {
			ClearLog();

		}
		private void btnImport_Click(object sender, EventArgs e) {
            try {
                Import();
            } catch(Exception ex) {
                DemoUtils.ReportOutlookError("import appointments from Microsoft Outlook", ex.Message);
            }
		}
		void Import() {
			AppointmentImporter importer = schedulerStorage.CreateOutlookImporter();
            ((ISupportCalendarFolders)importer).CalendarFolderName = CalendarName;
			importer.AppointmentImporting += new AppointmentImportingEventHandler(importer_AppointmentImporting);
			importer.AppointmentImported += new AppointmentImportedEventHandler(importer_AppointmentImported);
            importer.OnException += new ExchangeExceptionEventHandler(importer_OnException);

			Cursor oldCur = Cursor.Current;
			Cursor.Current = Cursors.WaitCursor;
			BeforeImport(importer.SourceObjectCount);
			try {
                importer.Import(System.IO.Stream.Null);
            }
			finally {
				AfterImport();
				Cursor.Current = oldCur;
			}
		}

        void importer_OnException(object sender, ExchangeExceptionEventArgs e) {
            string errText = e.OriginalException.Message;
            bool canTerminate = true;
            OutlookExchangeChangedOccurrenceExceptionEventArgs occurrenceArgs = e as OutlookExchangeChangedOccurrenceExceptionEventArgs;
            if (occurrenceArgs != null) {
                canTerminate = false;
                if (occurrenceArgs.OutlookAppointment != null && occurrenceArgs.Occurrence != null) {
                    errText += String.Format("\r\nChanged occurrence with index {0} is imported as normal occurrence '{1}' started on {2:D} at {3}\n", occurrenceArgs.Occurrence.RecurrenceIndex, occurrenceArgs.Occurrence.Subject,
                            occurrenceArgs.Occurrence.Start, occurrenceArgs.Occurrence.Start.TimeOfDay);
                }
            } else {
                OutlookExchangeExceptionEventArgs args = e as OutlookExchangeExceptionEventArgs;
                if (args != null) {
                    if (args.OutlookAppointment != null) {
                        errText += String.Format("\r\nEvent '{0}' started on {1:D} at {2}\n", args.OutlookAppointment.Subject,
                                args.OutlookAppointment.Start, args.OutlookAppointment.Start.TimeOfDay);
                    }
                }
            }

            AddToLog(errText);
            AppointmentImporter importer = (AppointmentImporter)sender;
               
            e.Handled = true;
            if (canTerminate) {
                importer.Terminate();
                throw e.OriginalException;
            }
        }
		void importer_AppointmentImporting(object sender, AppointmentImportingEventArgs e) {
			OutlookAppointmentImportingEventArgs args = (OutlookAppointmentImportingEventArgs)e;
			AddToLog(CreateImportLogString(args));
			
			ProgressBarControl1.Position++;
			ProgressBarControl1.Update();

			bool cancel = args.OutlookAppointment.IsRecurring ? IsCancelForRecurring : IsCancelForNonRecurring;
			if (cancel) {
				e.Cancel = true;
				AddToLog("Cancelled because of the appointment type");
			}
		}
		void importer_AppointmentImported(object sender, AppointmentImportedEventArgs e) {
			AddToLog(String.Format("Successfully imported: {0}", DemoUtils.FormatAppointmentString(e.Appointment)));
		}
	}
}
